les/pruner.go

```diff
@@ -0,0 +1,117 @@
+// Copyright 2019 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package les
+
+import (
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common/math"
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+// errPrunerClosed is returned if pruner is closed.
+var errPrunerClosed = errors.New("pruner is closed")
+
+// pruner is responsible for pruning historical light chain data.
+type pruner struct {
+	checkpoint uint64
+	closeCh    chan struct{}
+	addCh      chan *core.ChainIndexer
+	wg         sync.WaitGroup
+}
+
+// newPruner returns a light chain pruner instance.
+func newPruner(checkpoint uint64) *pruner {
+	pruner := &pruner{
+		checkpoint: checkpoint,
+		closeCh:    make(chan struct{}),
+		addCh:      make(chan *core.ChainIndexer),
+	}
+	pruner.wg.Add(1)
+	go pruner.loop()
+	return pruner
+}
+
+// register adds a batch of specific chain indexers as pruning target.
+func (p *pruner) register(indexers ...*core.ChainIndexer) error {
+	for _, indexer := range indexers {
+		select {
+		case p.addCh <- indexer:
+		case <-p.closeCh:
+			return errPrunerClosed
+		}
+	}
+	return nil
+}
+
+// close notifies all background goroutines belonging to pruner to exit.
+func (p *pruner) close() {
+	close(p.closeCh)
+	p.wg.Wait()
+}
+
+// loop periodically queries the status of chain indexers and prunes useless
+// historical chain data.
+func (p *pruner) loop() {
+	defer p.wg.Done()
+
+	var (
+		last     = p.checkpoint
```

Comment:
Does this mean that every time the client is started you begin deleting old chain data starting from the checkpoint? This feels a bit wasteful.

Reply:
Just an idea: you could change Prune to use a db iterator to find and delete everything in the db from the beginning to the given section instead of iterating with a for loop on every block number of the specific section. This would have multiple advantages:
- you do not need to remember what was pruned already
- it is efficient because chain data db keys are prefixed with block number (this was one of the reasons I did it like this)
- this way you can also prune cached chain data that was ODRed after pruning that section. For example log searching can download many old receipts and it is nice to cache them for a while but it would also be good to throw them away eventually. If you just clean everything before the new section every time a new section is processed then I think it is good enough.